kvbc/include/kvbc_adapter/kv_blockchain_adapter.hpp

```diff
@@ -0,0 +1,158 @@
+      kvbc_.emplace<CatBCimpl>(kv);
+    }
+  }
+
+  // Creates a key-value blockchain.
+  // If `category_types` is nullopt, the persisted categories in storage will be used.
+  // If `category_types` has a value, it should contain all persisted categories in storage at a minimum. New ones
+  // will be created and persisted. Users are required to pass a value for `category_types` on first construction
+  // (i.e. a new blockchain) in order to specify the categories in use. Failure to do so will generate an exception.
+  explicit KeyValueBlockchain(
+      const std::shared_ptr<concord::storage::rocksdb::NativeClient>& native_client,
+      bool link_st_chain,
+      const std::optional<std::map<std::string, categorization::CATEGORY_TYPE>>& category_types = std::nullopt) {
+    if (bftEngine::ReplicaConfig::instance().kvBlockchainVersion == BLOCKCHAIN_VERSION::CATEGORIZED_BLOCKCHAIN) {
+      auto kv = std::make_shared<CatBCimpl::element_type>(native_client, link_st_chain, category_types);
+      kvbc_.emplace<CatBCimpl>(kv);
+    }
+  }
+
+  // The noop converter returns the input string as is, without modifying it.
+  static const categorization::Converter kNoopConverter;
+
+  /////////////////////// Add Block ///////////////////////
+
+  BlockId addBlock(categorization::Updates&& updates);
+
+  /////////////////////// Delete block ///////////////////////
+
+  bool deleteBlock(const BlockId& blockId);
+  void deleteLastReachableBlock();
+
+  /////////////////////// Raw Blocks ///////////////////////
+
+  // Adds raw block and tries to link the state transfer blockchain to the main blockchain
+  // If linkSTChainFrom is true, try to remove blocks form the state transfer chain to the blockchain
+  void addRawBlock(const categorization::RawBlock& block, const BlockId& block_id, bool lastBlock = true);
+  std::optional<categorization::RawBlock> getRawBlock(const BlockId& block_id) const;
+
+  /////////////////////// Info ///////////////////////
+  BlockId getGenesisBlockId() const;
+  BlockId getLastReachableBlockId() const;
+  std::optional<BlockId> getLastStatetransferBlockId() const;
+
+  std::optional<categorization::Hash> parentDigest(BlockId block_id) const;
+  bool hasBlock(BlockId block_id) const;
+
+  std::vector<std::string> getStaleKeys(BlockId block_id,
+                                        const std::string& category_id,
+                                        const categorization::ImmutableOutput& updates_info) const;
+
+  std::vector<std::string> getStaleKeys(BlockId block_id,
+                                        const std::string& category_id,
+                                        const categorization::VersionedOutput& updates_info) const;
+
+  std::vector<std::string> getStaleKeys(BlockId block_id,
+                                        const std::string& category_id,
+                                        const categorization::BlockMerkleOutput& updates_info) const;
```

Comment:
Same as above, not part of an interface, and might not be supported by the v4 kvbc.